### DISCUSSION TITLE
Convert timestamp to my local time

### DISCUSSION THREAD

**Discussion by ngms17** [jertel/elastalert2#1148] [Q&A] [answered]
Noticed that the timestamp showed on the aggregation email are in UTC. My local time is UTC +1 and all my log are on that timestamp.

I tried the following enhancement:
```

class TimeEnhancement(BaseEnhancement):
    def process(self, match):
        for k, v in match.items():
            if isinstance(v, str) and v.endswith('Z'):
                try:
                    match[k] = pretty_ts(v)
                except:
                    pass
```

Although this converts to the correct timestamp, it converts in this format `2023-03-28 11:52 WEST`. I need this field to be in the default format `2023-03-28T09:05:26.795Z`.

Can someone help me?

**Comment by jertel** [comment]
Replace the pretty_ts statement with this: `match[k] = pretty_ts(v, True, '%Y-%m-%dT%H:%M:%S.%f')`. If you want the timezone added to the end you can append `%Z` or `%z` to the string.

**Reply by ngms17** [reply]
Thanks for the reply.

Unfortunately it still shows as "2023-03-29 09:43 WEST". The hour is correct but the format is still not in the format that i need. This is the Enhancement:

```
class TimeEnhancement(BaseEnhancement):
    def process(self, match):
        for k, v in match.items():
            if isinstance(v, str) and v.endswith('Z'):
                try:
                    match[k] = pretty_ts(v, True, '%Y-%m-%dT%H:%M:%S.%f')
                except:
                    pass
```

**Reply by jertel** [reply]
It doesn't seem to be using the new code then. Add some debug print statements to verify it's actually using the updated enhancement.

**Reply by ngms17** [reply]
Running `python3 -m elastalert.elastalert --verbose --rule /etc/elastalert2/rules/malware.yaml --config /etc/elastalert2/config.yaml` inside the /etc/elastalert2 directory gives me the following error:

`ImportError: cannot import name 'BaseEnhancement' from partially initialized module 'elastalert.enhancements' (most likely due to a circular import) (/etc/elastalert2/elastalert/enhancements.py)`

**Reply by ngms17** [reply]
I have atually tried to create a new enhancement file:

Tried creating a new directory with a new enhacement file but it does not do anything. File is create inside `/etc/elastalert2/elastalert_modules/my_enhancements.py`.

Then, in the rule config file, added: 

```
match_enhancements:
- "elastalert_modules.my_enhancements.MyEnhancement"
```

**Reply by jertel** [reply]
So to recap:
- You wrote an enhancement yesterday and it worked but you weren't happy with the date format.
- Today you changed the enhancement to force it to use my suggested format.
- Now the enhancement gives an error and "does not do anything"

Is this correct?

**Reply by ngms17** [reply]
Recap:

1. The first enhancement that i showed worked but not in the date format that i wanted. This enhancement was created in the default `enhancement.py` file located at `/etc/elastalert2/elastalert`.

2. I changed the enhancement to force it to use my suggested format. The enchancement worked but it was presenting the data format that i did not wanted.

3. Then, i tried to create a whole new enhancement file like presented in the documentation. Created a new directory `/etc/elastalert2/elastalert_modules` with a file called `my_enhancements.py` inside. Then created a new class `MyEnhancement` with the same suggested code that you gave me. That enhancement did not work at all, and the timestamp was in the correct format but not in my local time

4. Tried to reinstall elastalert2. Using the default TimeEnhancement, timestamp is in the correct timezone but not in the needed format. Timestamp shows as `2023-03-29 13:32 WEST`

```
class TimeEnhancement(BaseEnhancement):
      def process(self, match):
            match['@timestamp'] = pretty_ts(match['@timestamp'])
```

5. Change the TimeEnhancement code to the new code and still presents as `2023-03-29 13:32 WEST`. So this confirms that this enhancement is indeed working but not formatting the date

**Reply by jertel** [reply]
How did you install ElastAlert 2?

**Reply by ngms17** [reply]
```
cd /etc
git clone https://github.com/jertel/elastalert2.git
cd /etc/elastalert2
pip3 install -r requirements.txt
python3 setup.py install
```

Then created a elastalert2.service on /lib/systemd/system/elastalert2.service

```
[Unit]
Description=elastalert
After=multi-user.target

[Service]
Type=simple
WorkingDirectory=/etc/elastalert2
ExecStart=/usr/local/bin/elastalert --config /etc/elastalert2/config.yaml --verbose --rule /etc/elastalert2/rules/malware.yaml
StandardOutput=syslog
StandardError=syslog
KillSignal=SIGKILL
PIDFile=/var/run/elastalert2.pid

[Install]
WantedBy=multi-user.target
```

**Reply by jertel** [reply, marked as answer]
Using /etc as a working project directory is a bit unusual, so I was curious how that /etc/elastalert2 dir came to exist. I suspect the problem is that you are editing a file in the cloned repo, but ElastAlert2 is actually running from the installed wheel package under the /usr/local/lib/... tree.

**Reply by jertel** [reply]
I don't think it's fair to ask the community to "walk you through" the installation process when it is already documented, and has been successful to the general public for years. If we had a better way to explain it then we'd have already put it in the documentation. If there is something that you don't understand, please be specific with your question.
